chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -0,0 +1,490 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{btree_map::Entry, BTreeMap};
+
+use common::{
+    chain::{timelock::OutputTimeLock, AccountOutPoint, AccountSpending, ChainConfig, TxOutput},
+    primitives::{Amount, BlockDistance, BlockHeight},
+};
+use pos_accounting::PoSAccountingView;
+
+use crate::error::ConnectTransactionError;
+
+use super::IOPolicyError;
+
+/// `ConstrainedValueAccumulator` helps avoiding messy inputs/outputs combinations analysis by
+/// providing a set of properties that should be satisfied. For example instead of checking that
+/// all outputs are timelocked when the pool is decommissioned `ConstrainedValueAccumulator` gives a way
+/// to check that an accumulated output value is locked for sufficient amount of time which allows
+/// using other valid inputs and outputs in the same tx.
+///
+/// TODO: this struct can be extended to collect tokens replacing `AmountsMap`
+pub struct ConstrainedValueAccumulator {
+    unconstrained_value: Amount,
+    timelock_constrained: BTreeMap<BlockDistance, Amount>,
+}
+
+impl ConstrainedValueAccumulator {
+    pub fn new() -> Self {
+        Self {
+            unconstrained_value: Amount::ZERO,
+            timelock_constrained: Default::default(),
+        }
+    }
+
+    /// Return accumulated amounts that are left
+    // TODO: for now only used in tests but can be used to calculate fees
+    #[allow(dead_code)]
+    pub fn consume(self) -> Result<Amount, IOPolicyError> {
+        self.timelock_constrained
+            .values()
+            .copied()
+            .sum::<Option<Amount>>()
+            .and_then(|v| v + self.unconstrained_value)
+            .ok_or(IOPolicyError::AmountOverflow)
+    }
+
+    pub fn process_input_utxo(
+        &mut self,
+        chain_config: &ChainConfig,
+        block_height: BlockHeight,
+        pos_accounting_view: &impl PoSAccountingView,
+        output: &TxOutput,
+    ) -> Result<(), ConnectTransactionError> {
+        match output {
+            TxOutput::Transfer(value, _)
+            | TxOutput::LockThenTransfer(value, _, _)
```

Comment:
Once we want to use this struct for fees, keep in mind that we have to do arithmetic with timelock values to be able to spend anything.

Reply:
but we already have a separate mod `timelock_check` for that

Reply:
But we won't be able to calculate the valid timelocks if we don't do that. When the time comes we'll discuss it.